src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java

```diff
@@ -760,40 +765,60 @@ public boolean isKnownParticipant( UUID studyId, String participantId ) {
+        Set<UUID> result = new HashSet<>();
+
         Set<UUID> studyEntityKeyIds = studies
                 .stream()
                 .map( study -> UUID.fromString( study.get( ID_FQN ).iterator().next().toString() ) )
                 .collect( Collectors.toSet() );
 
-        Set<UUID> studyIds = studies
-                .stream()
-                .map( study -> UUID.fromString( study.get( STRING_ID_FQN ).iterator().next().toString() ) )
-                .collect( Collectors.toSet() );
+        //mapping from studyEKID -> studyId
+        Map<UUID, UUID> studyEntityKeyIdMap = new HashMap<>();
+
+        studies.forEach( study -> {
+            UUID studyId = UUID.fromString( study.get( STRING_ID_FQN ).iterator().next().toString() );
+            UUID entityKeyId = UUID.fromString( study.get( ID_FQN ).iterator().next().toString() );
+            studyEntityKeyIdMap.put( entityKeyId, studyId );
+        } );
 
         Map<UUID, List<NeighborEntityDetails>> studyNeighbors = searchApi
-                .executeEntityNeighborSearchBulk(
+                .executeFilteredEntityNeighborSearch(
                         studyESID,
-                        studyEntityKeyIds
+                        new EntityNeighborsFilter(
+                                studyEntityKeyIds,
+                                java.util.Optional.of( ImmutableSet.of( notificationsESID ) ),
+                                java.util.Optional.of( ImmutableSet.of( studyESID ) ),
+                                java.util.Optional.of( ImmutableSet.of( partOfESID ) )
+                        )
                 );
 
-        // studies with notifications enabled have 'ol.id' property in the corresponding associationDetails set to
-        // the value of the studyId
-        return studyNeighbors
-                .entrySet()
-                .stream()
-                .filter( entry -> entry
-                        .getValue()
-                        .stream()
-                        .filter( neighbor -> neighbor.getNeighborEntitySet().isPresent() &&
-                                neighbor.getNeighborEntitySet().get().getName()
-                                        .startsWith( NOTIFICATION_ENTITY_SET_PREFIX ) &&
-                                neighbor.getAssociationDetails().containsKey( OL_ID_FQN ) &&
-                                studyIds.contains( UUID
-                                        .fromString( neighbor.getAssociationDetails().get( OL_ID_FQN ).iterator().next()
-                                                .toString() ) ) )
-                        .count() != 0 )
-                .map( Map.Entry::getKey )
-                .collect( Collectors.toSet() );
+        /*
+         * studies with notifications enabled have 'ol.id' property in the corresponding
+         * associationDetails set to the value of the studyId
+         * For each study, there is can only be at most 1 notification -> partof -> study association,
```

Comment:
Isn't this gonna change very soon when there will be a notification for the questionnaire?

Reply:
um, good point. I actually haven't thought about how to do notifications for the questionnaire. Maybe we should have a `notification -> partof -> questionnaire` association?  I'd like to sync about that @jokedurnez